### PR TITLE
Read species size correctly

### PR DESCRIPTION
**Pull request by RevathiJambunathan**
For diagnostics, 
when the user specifies `diag_lo` and `diag_hi`, we read geometry filter and we also temporarily disable particle-io for slices (or diags over a small region of the domain).

When doing so, we loop over number of species, but this was not done correctly for the 2D vector and is fixed in this PR.
```
        for (int i_buffer = 0; i_buffer < m_num_buffers; ++i_buffer ) {
            for (int i = 0; i < m_output_species[i_buffer].size(); ++i) {
                m_output_species[i_buffer][i].m_do_geom_filter = true;
            }
         ...
        }
``
```

We now access species size with `m_output_species[i_buffer].size()` instead of `m_output_species.size()` which would return the number of buffers, instead of number of species in a given buffer.

Thanks to Andy Nonaka (@ajnonaka ) for finding this issue!

### CODE REVIEW

File: Source/Diagnostics/Diagnostics.cpp [hunk 1]
```diff
@@ -222,7 +222,7 @@ Diagnostics::InitData ()
         // Note that the filter is set for every ith snapshot, and the number of snapshots
         // for full diagnostics is 1, while for BTD it is user-defined.
         for (int i_buffer = 0; i_buffer < m_num_buffers; ++i_buffer ) {
-            for (int i = 0; i < m_output_species.size(); ++i) {
+            for (int i = 0; i < m_output_species[i_buffer].size(); ++i) {
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]: This might introduce the same bug as #2738 and needs some extra check?

Comment (RevathiJambunathan) [reply Source/Diagnostics/Diagnostics.cpp]: Ah I see. Yes! this is true!
Thanks for catching it!
WIll fix this

Comment (RevathiJambunathan) [reply Source/Diagnostics/Diagnostics.cpp]: The latest commit should fix it, I think

Do you have an input file that triggered this Backtrace in #2738 
I have tested it for a case without any species.

Comment (RevathiJambunathan) [reply Source/Diagnostics/Diagnostics.cpp]: Ah never mind. I found the input file, as was mentioned in #2738  .
I ran this case locally and it works now.

Comment (ax3l) [reply Source/Diagnostics/Diagnostics.cpp]: Thank you!

Can you please replace the unchecked `operator[]` with `.at()`? This would make this way safer, because of a runtime check.

I cannot fully understand why this part would work. `m_num_buffers` is not zero in this case, but the size of `m_output_species` is. The access of `m_output_species[0]` is then undefined behavior (a memory violation).

Comment (RevathiJambunathan) [reply Source/Diagnostics/Diagnostics.cpp]: Thank you @ax3l . I just realized why you mentioned this. 
Now when write_species is zero, species output is cleared for the ith buffer, but the whole 2D species_output vector is not cleared.

I have run this with DEBUG and also used` .at()`

**Comment by ax3l** [comment]
Thank you!

Does this PR also fix @TiberiusRheaume's problem in #2738? :)

File: Source/Diagnostics/Diagnostics.cpp [hunk 2]
```diff
@@ -214,33 +214,34 @@ Diagnostics::InitData ()
...
+        amrex::Vector <amrex::Real> dummy_val(AMREX_SPACEDIM);
+        if ( queryArrWithParser(pp_diag_name, "diag_lo", dummy_val, 0, AMREX_SPACEDIM) ||
+             queryArrWithParser(pp_diag_name, "diag_hi", dummy_val, 0, AMREX_SPACEDIM) ) {
+            // set geometry filter for particle-diags to true when the diagnostic domain-extent
+            // is specified by the user.
+            // Note that the filter is set for every ith snapshot, and the number of snapshots
+            // for full diagnostics is 1, while for BTD it is user-defined.
+            for (int i_buffer = 0; i_buffer < m_num_buffers; ++i_buffer ) {
+                for (int i = 0; i < m_output_species.at(i_buffer).size(); ++i) {
+                    m_output_species[i_buffer][i].m_do_geom_filter = true;
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]: Uses range-changed access for vector in vector:
```suggestion
                    m_output_species.at(i_buffer).at(i).m_do_geom_filter = true;
```

File: Source/Diagnostics/Diagnostics.cpp [hunk 3]
```diff
@@ -214,33 +214,34 @@ Diagnostics::InitData ()
...
+        if ( queryArrWithParser(pp_diag_name, "diag_lo", dummy_val, 0, AMREX_SPACEDIM) ||
+             queryArrWithParser(pp_diag_name, "diag_hi", dummy_val, 0, AMREX_SPACEDIM) ) {
+            // set geometry filter for particle-diags to true when the diagnostic domain-extent
+            // is specified by the user.
+            // Note that the filter is set for every ith snapshot, and the number of snapshots
+            // for full diagnostics is 1, while for BTD it is user-defined.
+            for (int i_buffer = 0; i_buffer < m_num_buffers; ++i_buffer ) {
+                for (int i = 0; i < m_output_species.at(i_buffer).size(); ++i) {
+                    m_output_species.at(i_buffer).at(i).m_do_geom_filter = true;
+                }
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]: Simple & compact:
```suggestion
                for (auto& v : m_output_species.at(i_buffer))
                    v.m_do_geom_filter = true;
```

File: Source/Diagnostics/Diagnostics.cpp [hunk 4]
```diff
@@ -214,33 +214,34 @@ Diagnostics::InitData ()
...
-
     if (write_species == 0) {
         if (m_format == "checkpoint"){
             amrex::Abort("For checkpoint format, write_species flag must be 1.");
         }
         // if user-defined value for write_species == 0, then clear species vector
-        m_output_species.clear();
+        for (int i_buffer = 0; i_buffer < m_num_buffers; ++i_buffer ) {
+            m_output_species.at(i_buffer).clear();
+        }
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp:224]:
```suggestion
        for (auto& v : m_output_species)
            m_output_species.clear();
```

File: Source/Diagnostics/Diagnostics.cpp [hunk 5]
```diff
@@ -214,33 +214,32 @@ Diagnostics::InitData ()
...
-    }
-
     if (write_species == 0) {
         if (m_format == "checkpoint"){
             amrex::Abort("For checkpoint format, write_species flag must be 1.");
         }
         // if user-defined value for write_species == 0, then clear species vector
-        m_output_species.clear();
+        for (auto& v : m_output_species)
+            m_output_species.clear();
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]: Reva on Slack:
>  I do not clear m_output_species but only the array for the ith  buffer
> This way every independent buffer can have different species (it is not supported now, but if it is in the future, then this is keeping it general and also avoiding the error)

Comment (ax3l) [reply Source/Diagnostics/Diagnostics.cpp]:
```suggestion
        for (int i_buffer = 0; i_buffer < m_num_buffers; ++i_buffer ) {
            m_output_species.at(i_buffer).clear();
        }
```

File: Source/Diagnostics/Diagnostics.cpp [hunk 6]
```diff
@@ -214,33 +214,33 @@ Diagnostics::InitData ()
...
-        amrex::Print() << " WARNING: For full diagnostics on a reduced domain, particle io is not supported, yet! Therefore, particle-io is disabled for this diag " << m_diag_name << "\n";
-    }
-
     if (write_species == 0) {
         if (m_format == "checkpoint"){
             amrex::Abort("For checkpoint format, write_species flag must be 1.");
         }
         // if user-defined value for write_species == 0, then clear species vector
-        m_output_species.clear();
+        for (int i_buffer = 0; i_buffer < m_num_buffers; ++i_buffer ) {
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp:222]: Reva on Slack:
>  I do not clear m_output_species but only the array for the ith  buffer
> This way every independent buffer can have different species (it is not supported now, but if it is in the future, then this is keeping it general and also avoiding the error)

I think we should use another, more appropriate data structure for this workflow when we need it in the future, that simplifies the book-keeping. :)
For instance: `std::map<>` of vectors for each species, etc.
I think that buffer index tracking is very error prone and hard to follow/extent.

Comment (RevathiJambunathan) [reply Source/Diagnostics/Diagnostics.cpp:222]: Should we do that in a separate PR?

**Comment by RevathiJambunathan** [comment]
@ax3l yes this fixes @TiberiusRheaume issue as well

File: Source/Diagnostics/Diagnostics.cpp [hunk 7]
```diff
@@ -219,9 +219,8 @@ Diagnostics::InitData ()
             amrex::Abort("For checkpoint format, write_species flag must be 1.");
         }
         // if user-defined value for write_species == 0, then clear species vector
-        for (int i_buffer = 0; i_buffer < m_num_buffers; ++i_buffer ) {
-            m_output_species.at(i_buffer).clear();
-        }
+        for (auto& v : m_output_species)
```
Comment (RevathiJambunathan) [inline Source/Diagnostics/Diagnostics.cpp]: Ah thanks for fixing this!

File: Source/Diagnostics/Diagnostics.cpp (hunk 6 above)
Comment (ax3l) [reply Source/Diagnostics/Diagnostics.cpp:222]: Yes, that would be a good refactoring in a separate PR after this fix is merged :)

File: Source/Diagnostics/Diagnostics.cpp [hunk 8]
```diff
@@ -214,33 +214,34 @@ Diagnostics::InitData ()
...
+            for (int i_buffer = 0; i_buffer < m_num_buffers; ++i_buffer ) {
+                for (auto& v : m_output_species.at(i_buffer)) {
+                    v.m_do_geom_filter = true;
+                }
+                // Disabling particle-io for reduced domain diagnostics by reducing
+                // the particle-diag vector to zero.
+                // This is a temporary fix until particle_buffer is supported in diagnostics.
+                m_output_species.at(i_buffer).clear();
+            }
+            amrex::Print() << " WARNING: For full diagnostics on a reduced domain, particle io is not supported, yet! Therefore, particle-io is disabled for this diag " << m_diag_name << "\n";
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.cpp]: Do you want to use the new warning logger for this? :)

https://warpx.readthedocs.io/en/latest/developers/warning_logger.html
```suggestion
            std::string warnMsg = "For full diagnostics on a reduced domain, particle I/O is not supported, yet! Therefore, particle I/O is disabled for this diagnostics: ";
            warnMsg += m_diag_name;
            WarpX::GetInstance().RecordWarning("Diagnostics", warnMsg);
```

**Review by ax3l — APPROVED** [review]
Thank you for fixing these issues :tada: :blowfish: :tropical_fish:

**Comment by RevathiJambunathan** [comment]
Thank you @ax3l for reviewing the PR and adding the warning logger and modernized for loops!

File: Source/Diagnostics/Diagnostics.cpp (hunk 6 above)
Comment (ax3l) [reply Source/Diagnostics/Diagnostics.cpp:222]: Clarified on zoom: I meant to write
```C++
// for every (BTD) buffer or the single full diag,
// clear species vector if write_species == 0
for (auto& v : m_output_species)
    v.clear();
```
and that one works - there are no offsets in `i_buffer` for full diags without BTD, `m_num_buffers` is simply `1` then.
